libethereum/State.cpp

```diff
@@ -42,40 +44,39 @@ State::State(State const& _s):
 
 OverlayDB State::openDB(fs::path const& _basePath, h256 const& _genesisHash, WithExisting _we)
 {
-    auto const path = _basePath.empty() ? db::databasePath() : _basePath;
-    auto const chainPath = path / fs::path(toHex(_genesisHash.ref().cropped(0, 4))) /
-                           fs::path(toString(c_databaseVersion));
-    auto const statePath = chainPath / fs::path("state");
-
     if (db::isDiskDatabase())
     {
+        db_paths::setDatabasePaths(_basePath, _genesisHash);
```

Comment:
You call this from here and in `BlockChain` class, so it will be called at least twice, won't it?
Also it makes it impossible to have two instantces of `State` pointing to different databases.

It should rather be called once somewhere at the higher level (like main function or `Client` initialization).

Reply:
@gumb0 For context, when would we ever have two `State `instances pointing to different databases?

Reply:
Maybe in tests, maybe nowhere, but it's one thing to set the global data once and then only read it in various places, but writing global data from various places just doesn't seem good practice - both from thread-safety perspective and from general complexity management perspective